feat(core): support a basic *ngFor example in Ivy

alxhub opened the pull request.

mhevery commented on packages/core/src/render3/definition_interfaces.ts:
````diff
@@ -146,3 +148,121 @@ export interface ComponentDefArgs<T> extends DirectiveDefArgs<T> {
+ * }
+ * ```
+ */
+export function defineComponent<T>(componentDefinition: ComponentDefArgs<T>): ComponentDef<T> {
````
Is this a bad merge? Why are we defining new `defineComponent` It is duplicate of `definition.ts`

mhevery commented on packages/core/src/render3/di.ts:
```diff
@@ -378,7 +390,22 @@ class ViewContainerRef implements viewEngine_ViewContainerRef {
   }
   insert(viewRef: viewEngine_ViewRef, index?: number|undefined): viewEngine_ViewRef {
-    throw notImplemented();
+    if (index == null) index = this._node.data.views.length;
```
could you add an assert that `index < this._node.data.views.length`?

alxhub replied:
Done.

mhevery commented on packages/core/src/render3/di.ts:
```diff
@@ -378,7 +390,22 @@ class ViewContainerRef implements viewEngine_ViewContainerRef {
+      // Increment the container view count.
+      this._node.data.dynamicViewCount++;
+
+      // Look for the parent node, which should be an LView, and increment its dynamic view count.
```
add assertion that parent view is LView

alxhub replied:
Done.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1074,6 +1104,22 @@ export function containerRefreshEnd(): void {
 
+function refreshDynamicChildren() {
+  for (let current = currentView.child; current !== null; current = current.next) {
+    if (current.dynamicViewCount === 0) {
```
I find this form more readable: (and it is shorter)
```
if (current.dynamicViewCount !== 0 && (current as ContaineState).views) {
  for(....) {}
}
```
  @

alxhub replied:
How would that work? `current` is the loop variable, it needs to be checked inside the loop.

I've combined the two checks into a single `if` though.

mhevery reviewed (commented):
Let's clean this up and get the tests to pass, and merge this. The PR is getting large. we should create follow up PRs for additional features.

trotyl commented:
Should the benchmark examples be updated as well? Current examples are using manual `for` loop which is not consistent with old ones.

mary-poppins commented:
You can preview bf67fbd at https://pr21430-bf67fbd.ngbuilds.io/.

mary-poppins commented:
You can preview bca239f at https://pr21430-bca239f.ngbuilds.io/.

mary-poppins commented:
You can preview 6b388c8 at https://pr21430-6b388c8.ngbuilds.io/.

mary-poppins commented:
You can preview c1436b6 at https://pr21430-c1436b6.ngbuilds.io/.

mary-poppins commented:
You can preview 77a6eac at https://pr21430-77a6eac.ngbuilds.io/.

mary-poppins commented:
You can preview 2ff8f61 at https://pr21430-2ff8f61.ngbuilds.io/.

mary-poppins commented:
You can preview 862cdd8 at https://pr21430-862cdd8.ngbuilds.io/.

mary-poppins commented:
You can preview 2744f78 at https://pr21430-2744f78.ngbuilds.io/.

mary-poppins commented:
You can preview 4162513 at https://pr21430-4162513.ngbuilds.io/.

mary-poppins commented:
You can preview fbb71d7 at https://pr21430-fbb71d7.ngbuilds.io/.

mary-poppins commented:
You can preview 884200e at https://pr21430-884200e.ngbuilds.io/.

mary-poppins commented:
You can preview f0d8a3a at https://pr21430-f0d8a3a.ngbuilds.io/.

mary-poppins commented:
You can preview 0985bf8 at https://pr21430-0985bf8.ngbuilds.io/.

mhevery commented on packages/core/src/render3/definition.ts:
```diff
@@ -50,8 +56,59 @@ export function defineComponent<T>(componentDefinition: ComponentDefArgs<T>): Co
+  [key: string]: any;
+};
+
+// TODO(alxhub): this is a hack because NgOnChangesFeature needs `type` which is no longer a part
```
Remove the comment since this is the correct way to do this.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -17,20 +17,35 @@ import {assertEqual, assertLessThan, assertNotEqual, assertNotNull} from './asse
+// handled via setter injection which is resolved at startup.
+
+// import {LQuery_, QueryList_} from './query';
+let LQuery_: LQueryConstructor;
```
per our discussion this breaks tree shaking.

mhevery commented on packages/core/src/render3/interfaces.ts:
```diff
@@ -0,0 +1,579 @@
+/**
```
this file looks like a bad merge. Can you rebase on latest

pkozlowski-opensource commented on packages/core/src/render3/query.ts:
```diff
@@ -152,24 +160,15 @@ function geIdxOfMatchingDirective(node: LNode, type: Type<any>): number|null {
+    //} else if (read === QueryReadStrategy.ViewContainerRef) {
+    //  return getOrCreateContainerRef(nodeInjector);
+    //} else if (read === QueryReadStrategy.TemplateRef) {
+    //  return getOrCreateTemplateRef(nodeInjector);
```
No need to leave commented code, just remove

alxhub committed:
fix(ivy): assertLessThan asserts the wrong thing

assertLessThan() actually does the opposite of what it advertises.
It's only through luck that existing asserts have not failed
before. This changes assertLessThan to actually assert that the
value is less than something.

alxhub committed:
refactor(ivy): avoid circular dep with query/di/instructions

To prepare for pending ngForOf work, the dep from instructions -> query
should be broken. This will enable a dep from di -> instructions while
avoiding a di -> instructions -> query -> di cycle.

Analyzing this cycle also uncovered another problem: the implementation
of query() breaks tree-shaking through a hard dependency on DI concepts
of TemplateRef, ElementRef, ViewContainerRef. This is fundamentally due
to how query() can query for those values without any configuration.

Instead, this fix introduces the concept by employing the strategy
pattern, and redefining QueryReadType to pass a function which will
return one of the above values. This strategy is then used for 'read'
instead of an enum in cases where special values should be read from
the DI system.

alxhub committed:
feat(ivy): support for the ngForOf directive, with tests

Implement NgOnChangesFeature, ViewContainerRef, TemplateRef,
and the renderEmbeddedTemplate instruction, and wire together the
pieces required for the ngForOf directive to work.

mary-poppins commented:
You can preview 5c4160a at https://pr21430-5c4160a.ngbuilds.io/.

mary-poppins commented:
You can preview fdc264c at https://pr21430-fdc264c.ngbuilds.io/.

mary-poppins commented:
You can preview ae10296 at https://pr21430-ae10296.ngbuilds.io/.

mary-poppins commented:
You can preview f79c99d at https://pr21430-f79c99d.ngbuilds.io/.

mhevery commented on packages/core/src/render3/interfaces/query.ts:
```diff
@@ -48,15 +49,10 @@ export interface LQuery {
-  ViewContainerRef = 1,
-  TemplateRef = 2,
-}
+export interface QueryReadStrategy<T> { run(injector: LInjector, node: LNode): T; }
```
So `QueryReadStrategy` is a public Type since it will show up in the instructions `.d.ts` which means that `LInjector` and `LNode` will become public type as well and they are meant to be private. So I propose you define the type like so:

```
export declare class QueryReadType {
  private query_read_type_brand: any;
}
```

and than just force cast your Strategy into this type. This way it will look opaque to the outside world.

alxhub replied:
Done, nice trick.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -337,68 +335,33 @@ export function getOrCreateNodeInjector(): LInjector {
- */
-export function injectViewContainerRef(): ViewContainerRef {
-  return getOrCreateContainerRef(getOrCreateNodeInjector());
+export function getOrCreateNodeInjectorForNode(node: LElementNode | LContainerNode): LInjector {
```
I don't think this function needs to be in `instructions.ts` since it does not access any of the special state. Can you move it back to `di.ts`?

alxhub replied:
Done.

mhevery commented on packages/core/src/render3/di.ts:
```diff
@@ -311,6 +348,27 @@ export function getOrCreateElementRef(di: LInjector): viewEngine_ElementRef {
   return di.elementRef || (di.elementRef = new ElementRef(di.node.native));
 }
 
+export const QUERY_READ_STRATEGY_TEMPLATE = new QueryReadStrategy_<viewEngine_TemplateRef<any>>(
```
remove `_STRATEGY` from the name and cast it to `QueryReadType`

pkozlowski-opensource replied:
Hmm, ideally we don't want to have "query" in the name anywhere here, as there is nothing specific to queries here - we basically have a function that reads a particular token from an injector. 

In the commit description @alxhub is saying:

> Analyzing this cycle also uncovered another problem: the implementation
of query() breaks tree-shaking through a hard dependency on DI concepts
of TemplateRef, ElementRef, ViewContainerRef. This is fundamentally due
to how query() can query for those values without any configuration.

If we introduce query concepts here in DI we are just changing an arrow direction, but still linking 2 concepts. Having said this the situation is better here as we don't add _imports_ to queries.

At the end of a day queries and DI are related... so maybe we should move those strategies to a separate file (`queries_di.ts` or sth)?

alxhub replied:
@pkozlowski-opensource These constants are DI functions at heart, but related to queries. They live in `di.ts` but declare themselves (in name and type) to be query-related.

This could be avoided by exporting them from DI with non-query-specific names and types, and re-exporting them with the appropriate query type from `index.ts` or `query.ts`. However, after discussion with Misko, I think it's fine to leave them where they are.

mhevery commented on packages/core/src/render3/di.ts:
```diff
@@ -311,6 +348,27 @@ export function getOrCreateElementRef(di: LInjector): viewEngine_ElementRef {
+export const QUERY_READ_STRATEGY_MIX =
+    new QueryReadStrategy_<any>((injector: LInjector, node: LNode) => {
+      ngDevMode && assertNodeOfPossibleTypes(node, LNodeFlags.Container, LNodeFlags.Element);
+      debugger;
```
remove `debugger`;

mhevery commented on packages/core/src/render3/query.ts:
```diff
@@ -53,7 +56,12 @@ export interface QueryPredicate<T> {
+  read: QueryReadStrategy<T>|null;
+
+  /**
+   * Indicates a strategy to read the default value for a node.
```
`strategy` -> `token` to be consistent with `read`

After more reading I don't think we need this field. Revert.

pkozlowski-opensource replied:
I agree. We need to either:
- pass "default strategy" explicitly as `read` (better tree-shaking but more generated code);
- keep `readDefaultInjectable` (less tree shaking but less generated code as well).

Personally I would vote for passing "default strategy" explicitly

mhevery commented on packages/core/src/render3/query.ts:
```diff
@@ -211,8 +208,11 @@ function add(predicate: QueryPredicate<any>| null, node: LNode) {
             // is local name pointing to a directive?
             if (directiveIdx > -1) {
               predicate.values.push(node.view.data[directiveIdx]);
+            } else if (predicate.defaultRead) {
```
Looking at this I don't think we need separate `defalutRead` I think it can be part of the normal strategy. We should keep it simple. Just move the whole else black into a strategy. (Let's chat tomorrow in person)

pkozlowski-opensource replied:
Agree, see https://github.com/angular/angular/pull/21430/commits/a2674d48999862e3bbdea792f3edca048e79d9d3#r162373282

Could you please ping me when chatting about this particular aspect, as I want to keep track of changes in this part of the code?

alxhub replied:
`defaultRead` is gone!

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1907,15 +1870,18 @@ function valueInData<T>(data: any[], index: number, value?: T): T {
   return queryList;
 }
 
+export function queryRefresh(query: QueryList<any>): boolean {
```
I think we should revert this (put it back) since we should only keep instructions which need access to the global state here.

alxhub replied:
Done.

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -306,6 +309,34 @@ export function renderTemplate<T>(
+    let cm: boolean = false;
+    if (viewNode == null) {
+      const view = createLView(-1, renderer, {data: []});
+      view.template = template;
```
can template and context be optional args to `createLView`?

alxhub replied:
Done.

pkozlowski-opensource commented on packages/core/src/render3/di.ts:
```diff
@@ -311,6 +348,27 @@ export function getOrCreateElementRef(di: LInjector): viewEngine_ElementRef {
+export const QUERY_READ_STRATEGY_ELEMENT = new QueryReadStrategy_<viewEngine_ElementRef>(
+    (injector: LInjector) => getOrCreateElementRef(injector));
+
+export const QUERY_READ_STRATEGY_MIX =
```
Could we replace "mix" with something else? I guess it corresponds to the `readDefaultInjectable` function? How about `QUERY_READ_ELEMENT_DEFAULT` or sth similar?

alxhub replied:
Renamed to `QUERY_READ_FROM_NODE`.

pkozlowski-opensource commented:
@alxhub I left some comments in the code, mostly echoing items spotted by @mhevery already. As Misko, I would like to see "default read strategy" going away.

The only "bigger"  concern is connected with:

> To prepare for pending ngForOf work, the dep from query -> di needs
to be broken. This will enable a dep from di -> instructions while
avoiding a di -> instructions -> query -> di cycle.

In my mind we shouldn't have `di -> instructions` dependency, why do we need to break it down?
Please note that I was "fighting hard" to remove `di -> instructions` dependency previously here: https://github.com/mhevery/angular/commit/bbdea96a6679c84766fdb6cf66ae160b84523bc6

mary-poppins commented:
You can preview cf05bca at https://pr21430-cf05bca.ngbuilds.io/.

alxhub commented on packages/core/test/render3/query_spec.ts:
```diff
@@ -91,7 +93,7 @@ describe('query', () => {
         let tmp: any;
         if (cm) {
-          m(0, Q(Child, false, QueryReadType.ElementRef));
+          m(0, Q(Child, false, QUERY_READ_STRATEGY_ELEMENT));
```
Done.

alxhub commented on packages/core/src/render3/query.ts:
```diff
@@ -16,15 +16,18 @@ import {TemplateRef as viewEngine_TemplateRef} from '../linker/template_ref';
 
 const unusedValueToPlacateAjd = unused1 + unused2 + unused3 + unused4;
 
+export class QueryReadStrategy_<T> {
```
Renamed to `ReadFromInjectorFn`.

mary-poppins commented:
You can preview 4cdfdcb at https://pr21430-4cdfdcb.ngbuilds.io/.

mary-poppins commented:
You can preview 6f74445 at https://pr21430-6f74445.ngbuilds.io/.

alxhub commented on packages/core/src/render3/index.ts:
```diff
@@ -59,17 +61,16 @@ export {
+
 // clang-format on
-export {QueryList} from './query';
+export {QueryList, query as Q, queryRefresh as qR,} from './query';
```
Reformat as above.

alxhub replied:
Done.

alxhub commented on packages/core/src/render3/interfaces/query.ts:
```diff
@@ -48,15 +49,10 @@ export interface LQuery {
   track<T>(
       queryList: QueryList<T>, predicate: Type<any>|string[], descend?: boolean,
-      read?: QueryReadType|Type<T>): void;
+      read?: QueryReadType<T>|Type<T>, defaultRead?: QueryReadType<T>): void;
```
Remove from private interface.

alxhub replied:
Done.

alxhub commented on packages/core/src/render3/query.ts:
```diff
@@ -202,18 +204,13 @@ function add(predicate: QueryPredicate<any>| null, node: LNode) {
-            const requestedRead = readFromNodeInjector(nodeInjector, node, predicate.read);
-            if (requestedRead !== null) {
-              predicate.values.push(requestedRead);
+          if (predicate.read) {
```
!== null

alxhub replied:
Done.

alxhub commented on packages/core/src/render3/instructions.ts:
```diff
@@ -187,7 +189,8 @@ export function createLView(viewId: number, renderer: Renderer3, tView: TView):
     bindingStartIndex: null,
     creationMode: true,
-    viewHookStartIndex: null
+    viewHookStartIndex: null, template, context,
```
Explicitly assign here.

alxhub replied:
Done.

mary-poppins commented:
You can preview 1800df7 at https://pr21430-1800df7.ngbuilds.io/.

alxhub commented:
All changes in, should be good to go.

mary-poppins commented:
You can preview a3be5b3 at https://pr21430-a3be5b3.ngbuilds.io/.

mary-poppins commented:
You can preview 7ffe908 at https://pr21430-7ffe908.ngbuilds.io/.

mary-poppins commented:
You can preview 2185e6a at https://pr21430-2185e6a.ngbuilds.io/.

alxhub commented:
TAP failures are unrelated (closure error related to some project including analytics externs twice).
